### PR TITLE
Issue #12200: fix JavadocMethodCheck for records

### LINKED ISSUE #12200: False negative from JavadocMethod when a record canonical constructor is documented inside an interface
I have read check documentation: https://checkstyle.org/config_javadoc.html#JavadocMethod
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
$ javac Interface.java

$ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocMethod"/>
  </module>
</module>

$ cat Interface.java
package my.pack;

import static java.util.Objects.requireNonNull;

/** My outer interface. */
public interface Interface {

  /** My inner record. */
  public record Record(int field) {
    /**
     * Constructor.
     *
     * @param field Some field.
     */
    public Record {
      requireNonNull(field);
    }
  }
}

$ RUN_LOCALE="-Duser.language=en -Duser.country=US"
$ java $RUN_LOCALE -jar checkstyle-10.3.3-all.jar -c config.xml Interface.java
Starting audit...
[ERROR] C:\Users\jesper\tmp\checkstyle_bug\Interface.java:13: Unused Javadoc tag. [JavadocMethod]
Audit done.
Checkstyle ends with 1 errors.

$ cat OtherRecord.java
package my.pack;

import static java.util.Objects.requireNonNull;

/** My other record. */
public record OtherRecord(int field) {
  /**
   * Constructor.
   *
   * @param field Some field.
   */
  public OtherRecord {
    requireNonNull(field);
  }
}

$ java $RUN_LOCALE -jar checkstyle-10.3.3-all.jar -c config.xml OtherRecord.java
Starting audit...
Audit done.
```

---

I expect checkstyle to accept the javadoc for the record constructor - just it does when the record is not placed inside an interface.

### PR DESCRIPTION
**Pull request by strkkk**
Closes #12200

Diff Regression config: https://gist.githubusercontent.com/strkkk/238b6c97492ff6ffdb3be1603deca8fc/raw/fdba9e679917c42e7a2ba8a1b03e70073b8c4e05/config_single_2.xml

### CODE REVIEW

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodRecordsAndCompactCtors.java [hunk 1]
```diff
@@ -62,7 +62,7 @@ public record MySecondRecord() {
     }
 
     // Record component, compact ctor
-    public record MyThirdRecord(String myString) {
+    public record MyThirdRecord(String myString) { // violation
```
Comment (strkkk) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodRecordsAndCompactCtors.java:65]: new violation was discovered as well

**Comment by strkkk** [comment]
Github, generate report

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodRecordsAndCompactCtors.java (hunk 1 above)
Comment (romani) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodRecordsAndCompactCtors.java:65]: Please put violation message in comment, we demand this for all new changes.

Comment (romani) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodRecordsAndCompactCtors.java:65]: Remove trailing comment from
`@param properties some value // violation`

Comment (strkkk) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodRecordsAndCompactCtors.java]: done

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java [hunk 2]
```diff
@@ -708,11 +712,34 @@ private static List<JavadocTag> getMultilineNoArgTags(final Matcher noargMultili
...
+        return collectParamNames(params, TokenTypes.PARAMETER_DEF);
+    }
+
+    /**
+     * Computes the parameter nodes for a record constructor.
+     *
+     * @param ast the compact constructor node.
+     * @return the list of parameter nodes for ast.
+     */
+    private static List<DetailAST> getRecordParameters(DetailAST ast) {
```
[the remark this answers is not in the document]
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java]: done

**Review by romani — APPROVED** [review]
Ok to merge

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodRecordsAndCompactCtors.java [hunk 3]
```diff
@@ -62,7 +62,7 @@ public record MySecondRecord() {
     }
 
     // Record component, compact ctor
-    public record MyThirdRecord(String myString) {
+    public record MyThirdRecord(String myString) { // violation 'Expected @param tag for 'myString''
```
Comment (rnveach) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodRecordsAndCompactCtors.java:65]: Are these type of violations new? It doesn't really make sense to me for it be here where this is not the Javadoc, and Javadoc is on the constructor.

Comment (rnveach) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodRecordsAndCompactCtors.java:65]: Why is there no similar violation on `MyFourthRecord`?

**Review by pbludov — CHANGES_REQUESTED** [review]

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodRecordsAndCompactCtors.java (hunk 3 above)
Comment (rnveach) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodRecordsAndCompactCtors.java:65]: @pbludov I am still not seeing an answer to this on `MyFourthRecord`.

**Comment by strkkk** [comment]
I do not have time to finish this PR

**Pull request closed without merging**